### PR TITLE
refactor(runner): update Pandas and Default DataContainer

### PR DESCRIPTION
**Pull request by larme**
## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/runner/container.py [hunk 1]
```diff
@@ -96,19 +96,19 @@ def batches_to_batch(
         indices = list(
             itertools.accumulate(subbatch.shape[batch_dim] for subbatch in batches)
         )
-        indices.pop()
+        indices = [0] + indices
```
Comment (larme) [inline bentoml/_internal/runner/container.py:99]: I made this changes so every container type has consistent indices format returned

File: bentoml/_internal/runner/container.py [hunk 2]
```diff
@@ -148,13 +148,19 @@ def from_payload(  # pylint: disable=arguments-differ
...
         plasma_db: "ext.PlasmaClient" = Provide[DeploymentContainer.plasma_db],
     ) -> t.List[Payload]:
 
         batches = cls.batch_to_batches(batch, indices, batch_dim)
-        payloads = [cls.to_payload(subbatch, plasma_db) for subbatch in batches]
+
+        def to_payload(subbatch: "ext.NpNDArray"):
+            payload = cls.to_payload(subbatch, plasma_db)
+            payload.meta["batch_size"] = subbatch.shape[batch_dim]
+            return payload
```
Comment (larme) [inline bentoml/_internal/runner/container.py:161]: added batch_size in metadata of payload

File: bentoml/_internal/runner/container.py [hunk 3]
```diff
@@ -239,42 +242,101 @@ def payload_to_single(  # pylint: disable=arguments-differ
...
+    ) -> t.Tuple[t.List[t.Any], t.List[int]]:
         assert (
-            batch_axis == 0
-        ), "Default Runner DataContainer does not support batch_axies other than 0"
-        return list(singles)
+            batch_dim == 0
+        ), "Default Runner DataContainer does not support batch_dim other than 0"
+        batch = []
+        for subbatch in batches:
+            batch.extend(subbatch)
```
Comment (larme) [inline bentoml/_internal/runner/container.py:287]: I made some rough benchmark and it seems that for large amount of list, using `extend` is most efficient way. Any other suggestion?

File: bentoml/_internal/runner/container.py (hunk 1 above)
Comment (parano) [reply bentoml/_internal/runner/container.py:101]: don't we still need the `indices.pop` here? cc @bojiang @sauyon

Comment (parano) [reply bentoml/_internal/runner/container.py:101]: don't we still need the `indices.pop` here? cc @bojiang @sauyon

File: bentoml/_internal/runner/container.py [hunk 4]
```diff
@@ -241,42 +244,101 @@ def payload_to_single(  # pylint: disable=arguments-differ
...
+        indices: t.Sequence[int],
+        batch_dim: int = 0,
+        plasma_db: "ext.PlasmaClient" = Provide[DeploymentContainer.plasma_db],
+    ) -> t.List[Payload]:
+
+        batches = cls.batch_to_batches(batch, indices, batch_dim)
+
+        def to_payload(subbatch: "ext.PdDataFrame"):
+            payload = cls.to_payload(subbatch, plasma_db)
+            payload.meta["batch_size"] = subbatch.shape[batch_dim]
```
Comment (parano) [inline bentoml/_internal/runner/container.py:261]: question, do we use this `batch_size` anywhere?

Comment (parano) [inline bentoml/_internal/runner/container.py:261]: question, do we use this `batch_size` anywhere?

File: bentoml/_internal/runner/container.py [hunk 5]
```diff
@@ -241,42 +244,101 @@ def payload_to_single(  # pylint: disable=arguments-differ
...
+        cls,
+        payloads: t.Sequence[Payload],
+        batch_dim: int = 0,
+        plasma_db: "ext.PlasmaClient" = Provide[DeploymentContainer.plasma_db],
+    ) -> t.Tuple["ext.PdDataFrame", t.List[int]]:
+        batches = [cls.from_payload(payload, plasma_db) for payload in payloads]
+        return cls.batches_to_batch(batches, batch_dim)
 
 
 class DefaultContainer(DataContainer[t.Any, t.List[t.Any]]):
```
Comment (parano) [inline bentoml/_internal/runner/container.py:279]: Maybe we can rename this to Sequence container now?

Comment (parano) [inline bentoml/_internal/runner/container.py:279]: Maybe we can rename this to Sequence container now?

File: bentoml/_internal/runner/container.py [hunk 6]
```diff
@@ -175,42 +181,39 @@ class PandasDataFrameContainer(
...
     @classmethod
-    def singles_to_batch(
+    def batches_to_batch(
         cls,
-        singles: t.Sequence[t.Union["ext.PdDataFrame", "ext.PdSeries"]],
-        batch_axis: int = 0,
-    ) -> "ext.PdDataFrame":
+        batches: t.Sequence["ext.PdDataFrame"],
+        batch_dim: int = 0,
+    ) -> t.Tuple["ext.PdDataFrame", t.List[int]]:
```
Comment (parano) [inline bentoml/_internal/runner/container.py:188]: let's make a new type `BatchIndices` for `t.List[int]` here?

Comment (parano) [inline bentoml/_internal/runner/container.py:188]: let's make a new type `BatchIndices` for `t.List[int]` here?

**Comment by parano** [comment]
LGTM @larme - a few minor comments, let's address in follow up PR

**Pull request merged**